### PR TITLE
Add status resource for chart controller

### PR DESCRIPTION
**Pull request by rossf7**
Towards https://github.com/giantswarm/giantswarm/issues/4949

- Adds the status resource for the new chart CRD controller.
- For chartconfig CRs we only set the Helm Release status.
- For chart CRs we will set 4 fields.
- This is so they can be set in the app CR status and we can have funky metrics in app-operator.

### Fields

- AppVersion - e.g. Ingress Controller has a CVE how many clusters are affected.
- LastDeployed - the time the Helm Release was last deployed.
- Status - the Helm Release status.
- Version - the version of the Helm Release.

### CODE REVIEW

File: service/controller/chart/v1/resource/status/create.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+		// Return early. We will retry on the next execution.
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	currentStatus := v1alpha1.ChartStatus{
+		AppVersion:   releaseHistory.AppVersion,
+		Status:       releaseContent.Status,
+		LastDeployed: v1alpha1.DeepCopyTime{releaseHistory.LastDeployed},
```
Comment (rossf7) [inline service/controller/chart/v1/resource/status/create.go]: This is the deep copy magic we need for time.Time fields.

File: service/controller/chart/v1/resource/status/create.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	releaseName := key.ReleaseName(cr)
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("getting status for release %#q", releaseName))
+
+	releaseContent, err := r.helmClient.GetReleaseContent(ctx, releaseName)
+	if helmclient.IsReleaseNotFound(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("release %#q not found", releaseName))
```
Comment (xh3b4sd) [inline service/controller/chart/v1/resource/status/create.go]: What about this? I think this aligns more with the fact gathering primitives described in https://github.com/giantswarm/fmt/blob/master/go/log_messages.md#ensuring-state. 

```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not get status for release %#q", releaseName))
		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("release %#q not found", releaseName))
		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
```

File: service/controller/chart/v1/resource/status/create.go (hunk 1 above)
Comment (xh3b4sd) [reply service/controller/chart/v1/resource/status/create.go]: Ordering would be grand.

File: service/controller/chart/v1/resource/status/create.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		LastDeployed: v1alpha1.DeepCopyTime{releaseHistory.LastDeployed},
+		Version:      releaseHistory.Version,
+	}
+
+	if !Equals(currentStatus, key.ChartStatus(cr)) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("setting status for release %#q status to %#q", releaseName, releaseContent.Status))
+
+		crCopy := cr.DeepCopy()
+		crCopy.Status = currentStatus
+		_, err := r.g8sClient.ApplicationV1alpha1().Charts(cr.Namespace).UpdateStatus(crCopy)
```
Comment (xh3b4sd) [inline service/controller/chart/v1/resource/status/create.go]: I think we have to fetch the latest resource version here beforehand.

File: service/controller/chart/v1/resource/status/create.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("status set for release %#q", releaseName))
```
Comment (xh3b4sd) [inline service/controller/chart/v1/resource/status/create.go]:
```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("set status for release %#q", releaseName))
```

File: service/controller/chart/v1/resource_set.go [hunk 5]
```diff
@@ -73,8 +79,23 @@ func NewResourceSet(config ResourceSetConfig) (*controller.ResourceSet, error) {
...
+
+		statusResource, err = status.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
 	resources := []controller.Resource{
 		releaseResource,
+		statusResource,
```
Comment (xh3b4sd) [inline service/controller/chart/v1/resource_set.go]: Just wondering, why not executing it first?

File: service/controller/chart/v1/resource/status/resource.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	return r, nil
+}
+
+func (r *Resource) Name() string {
+	return Name
+}
+
+// Equals asseses the equality of ChartStatuses with regards to distinguishing
+// fields.
+func Equals(a, b v1alpha1.ChartStatus) bool {
```
Comment (kopiczko) [inline service/controller/chart/v1/resource/status/resource.go]: This and IsEmpty shouldn't be exported.

**Comment by rossf7** [comment]
> When Pawel is happy I am happy.

@kopiczko PTAL when you get time.

File: service/controller/chart/v1/resource/release/resource.go [hunk 7]
```diff
@@ -78,8 +78,8 @@ func (r *Resource) Name() string {
 	return Name
 }
 
-// Equals asseses the equality of ReleaseStates with regards to distinguishing fields.
-func (a *ReleaseState) Equals(b ReleaseState) bool {
+// equals asseses the equality of ReleaseStates with regards to distinguishing fields.
+func (a *ReleaseState) equals(b ReleaseState) bool {
```
Comment (kopiczko) [inline service/controller/chart/v1/resource/release/resource.go]: I'm not a fan of adding methods to data structs in general.

When it was a function it shouldn't be exported. When this is now a method I believe it should exported. I think private methods of struct should be used only by the struct instance itself (there are exceptions when the struct is a part of the public API).

This is my PoV. Not a blocker.

Comment (rossf7) [reply service/controller/chart/v1/resource/release/resource.go]: Ack. I converted the methods into functions. I think this is nicer and it aligns the Release and Status resources.

**Comment by rossf7** [comment]
Changes made. Going with this.

**Pull request merged**